### PR TITLE
MINIFICPP-2056 CWEL: support UserData in simple/flattened json

### PR DESCRIPTION
**Pull request by szaszm**
UserData in windows events is an unrestricted XML element, that may contain additional info, which is not part of the well-known tags under System and EventData. I've added a generic xml -> json function that covers the majority of use cases, and generates a simple json output (as opposed to the generic but convoluted raw format), but skips attributes on leaf nodes, and may have other limitations. We can make it smarter later if anyone encounters an event that is not properly converted by the current code.

----

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/windows-event-log/wel/JSONUtils.cpp [hunk 1]
```diff
@@ -162,9 +186,4 @@ std::string jsonToString(const rapidjson::Document& doc) {
   doc.Accept(writer);
   return buffer.GetString();
 }
-
-}  // namespace wel
-}  // namespace minifi
-}  // namespace nifi
-}  // namespace apache
-}  // namespace org
+} // namespace org::apache::nifi::minifi::wel
```
Comment (martinzink) [inline extensions/windows-event-log/wel/JSONUtils.cpp:189]: I think the linter failure comes from here `JSONUtils.cpp:189:  At least two spaces is best between code and comments  [whitespace/comments] [2]`

Comment (szaszm) [reply extensions/windows-event-log/wel/JSONUtils.cpp]: thanks, fixed

File: extensions/windows-event-log/wel/JSONUtils.cpp [hunk 2]
```diff
@@ -67,6 +59,29 @@ rapidjson::Value xmlDocumentToJSON(const pugi::xml_node& node, rapidjson::Docume
   return children;
 }
 
+void simplifiedGenericXmlToJson(const pugi::xml_node& node, rapidjson::Value& val, rapidjson::Document& doc, bool flatten = false) {
```
Comment (fgerlits) [inline extensions/windows-event-log/wel/JSONUtils.cpp:62]: these parameters could have better names, e.g `const pugi::xml_node& source, rapidjson::Value& target, rapidjson::Document& allocator_provider`

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp [hunk 3]
```diff
@@ -469,4 +470,65 @@ TEST_CASE("ConsumeWindowsEventLog batch commit size works", "[onTrigger]") {
...
+      <RuleSddl>D:(XA;;FX;;;S-1-1-0;(APPID://PATH Contains &quot;*&quot;))</RuleSddl>
+      <TargetUser>S-1-1-0</TargetUser>
+      <TargetProcessId>1234</TargetProcessId>
+      <FilePathLength>22</FilePathLength>
+      <FilePath>%SYSTEM32%\CSCRIPT.EXE</FilePath>
+      <FileHashLength>0</FileHashLength>
+      <FileHash></FileHash>
+      <FqbnLength>1</FqbnLength>
+      <Fqbn>-</Fqbn>
+    </RuleAndFileData>
```
Comment (fgerlits) [inline extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp:516]: can you add some nodes with attributes inside `UserData` to check & document what happens to `<Parent foo="bar"><Child/></Parent>` and `<Leaf foo="bar"></Leaf>` (maybe also `<AltLeaf foo="bar"/>`), please?

**Review by lordgamez — APPROVED** [review]
LGTM, just a single nitpick

File: extensions/windows-event-log/wel/JSONUtils.cpp (hunk 2 above)
Comment (szaszm) [reply extensions/windows-event-log/wel/JSONUtils.cpp]: Renamed them. Not exactly to your suggestions, but I'm pretty sure there should no longer be any confusion about their purpose. [66ab7e5](https://github.com/apache/nifi-minifi-cpp/pull/1515/commits/66ab7e50bb0de7428b0cd5101e9189431c0119ac)

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp [hunk 4]
```diff
@@ -469,4 +470,65 @@ TEST_CASE("ConsumeWindowsEventLog batch commit size works", "[onTrigger]") {
...
+    CHECK(expected_json == simple_json);
+  }
+  SECTION("flattened") {
+    const auto flattened_json = jsonToString(toFlattenedJSON(doc));
+    const auto expected_json = R"json({"Name":"Microsoft-Windows-AppLocker","Guid":"CBDA4DBF-8D5D-4F69-9578-BE14AA540D22","EventID":"8002","Version":"0","Level":"4","Task":"0","Opcode":"0","Keywords":"0x8000000000000000","SystemTime":"2023-02-06T16:58:09.008534Z","EventRecordID":"46","ProcessID":"1234","ThreadID":"1235","Channel":"Microsoft-Windows-AppLocker/EXE and DLL","Computer":"example.local","EventData":[],"PolicyNameLength":"3","PolicyName":"EXE","RuleNameLength":"9","RuleName":"All files","RuleSddlLength":"48","RuleSddl":"D:(XA;;FX;;;S-1-1-0;(APPID://PATH Contains \"*\"))","TargetUser":"S-1-1-0","TargetProcessId":"1234","FilePathLength":"22","FilePath":"%SYSTEM32%\\CSCRIPT.EXE","FileHashLength":"0","FileHash":"","FqbnLength":"1","Fqbn":"-"})json";  // NOLINT: long raw string, impractical to split
```
[the remark this answers is not in the document]
Comment (szaszm) [reply extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp]: added whitespace/line_length in [66ab7e5](https://github.com/apache/nifi-minifi-cpp/pull/1515/commits/66ab7e50bb0de7428b0cd5101e9189431c0119ac)

**Pull request closed without merging**